clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -0,0 +1,191 @@
+package com.linkedin.venice.fastclient.transport;
+
+import com.google.protobuf.ByteString;
+import com.linkedin.venice.client.store.transport.TransportClient;
+import com.linkedin.venice.client.store.transport.TransportClientResponse;
+import com.linkedin.venice.compression.CompressionStrategy;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.protocols.VeniceClientRequest;
+import com.linkedin.venice.protocols.VeniceReadServiceGrpc;
+import com.linkedin.venice.protocols.VeniceServerResponse;
+import com.linkedin.venice.request.RequestHelper;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class GrpcTransportClient extends InternalTransportClient {
+  private static final Logger LOGGER = LogManager.getLogger(GrpcTransportClient.class);
+  private final TransportClient r2TransportClient; // used for metadata requests
+
+  // field has public visibility for unit testing
+  public final Map<String, ManagedChannel> serverGrpcChannels;
+  final Map<String, String> veniceAddressToGrpcAddress;
+
+  public GrpcTransportClient(Map<String, String> nettyServerToGrpc, TransportClient r2TransportClient) {
+    serverGrpcChannels = new HashMap<>();
+    veniceAddressToGrpcAddress = new HashMap<>();
+    this.r2TransportClient = r2TransportClient;
+
+    for (Map.Entry<String, String> entry: nettyServerToGrpc.entrySet()) {
+      String nettyServer = entry.getKey().split(":")[0];
+      String grpcServer = nettyServer + ":" + entry.getValue();
+
+      veniceAddressToGrpcAddress.put(entry.getKey(), grpcServer);
+    }
+  }
+
+  public ManagedChannel getChannel(String requestPath) {
+    String portKey = requestPath.split("/")[2];
+
+    if (!veniceAddressToGrpcAddress.containsKey(portKey)) {
+      throw new VeniceException("No grpc server found for port: " + portKey);
+    }
+
+    return serverGrpcChannels.computeIfAbsent(
+        portKey,
+        k -> ManagedChannelBuilder.forTarget(veniceAddressToGrpcAddress.get(k)).usePlaintext().build());
+  }
+
+  @Override
+  public CompletableFuture<TransportClientResponse> get(String requestPath, Map<String, String> headers) {
+    String[] requestParts = RequestHelper.getRequestParts(requestPath);
+
+    if (!requestParts[1].equals("storage")) {
+      LOGGER.debug("Requesting metadata from R2 client");
+      return r2TransportClient.get(requestPath, headers);
+    }
+
+    // Lazily create channel for each venice server
+    ManagedChannel channel = getChannel(requestPath);
+
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
```

Comment:
Is it possible to do a lightweight mutation of the request path, such as replacing the first part by some different sub-path?

Ideally, we want to reuse the same request path..
Another way would be to have a special header in `headers` map to indicate whether this is a metadata request or not, and we can reuse the same request path between Netty request and GPRC request.

Reply:
I will look into this.